mpa/modules/models/segmentors/mean_teacher_segmentor.py

```diff
@@ -0,0 +1,102 @@
+from mmseg.ops import resize
+import torch
+import functools
+from collections import OrderedDict
+from mpa.utils.logger import get_logger
+logger = get_logger()
+
+
+@SEGMENTORS.register_module()
+class MeanTeacherNaive(BaseSegmentor):
+    def __init__(self, orig_type=None, unsup_weight=0.1, warmup_start_iter=30, **kwargs):
+        logger.info('MeanTeacherNaive Segmentor init!')
+        super(MeanTeacherNaive, self).__init__()
+        self.test_cfg = kwargs['test_cfg']
+        self.warmup_start_iter = warmup_start_iter
+        self.count_iter = 0
+
+        cfg = kwargs.copy()
+        if orig_type == 'EncoderDecoder':
+            cfg['type'] = 'EncoderDecoder'
+            self.align_corners = cfg['decode_head'].align_corners
+        self.model_s = build_segmentor(cfg)
+        self.model_t = build_segmentor(cfg)
+        self.unsup_weight = unsup_weight
+
+        # Hooks for super_type transparent weight load/save
+        self._register_state_dict_hook(self.state_dict_hook)
+        self._register_load_state_dict_pre_hook(
+            functools.partial(self.load_state_dict_pre_hook, self)
+        )
+
+    def extract_feat(self, imgs):
+        return self.model_s.extract_feat(imgs)
+
+    def simple_test(self, img, img_metas, **kwargs):
+        return self.model_s.simple_test(img, img_metas, **kwargs)
+
+    def aug_test(self, imgs, img_metas, **kwargs):
+        return self.model_s.aug_test(imgs, img_metas, **kwargs)
+
+    def forward_dummy(self, img, **kwargs):
+        return self.model_s.forward_dummy(img, **kwargs)
+
+    def forward_train(self, img, img_metas, gt_semantic_seg, **kwargs):
+        self.count_iter += 1
+        if self.warmup_start_iter > self.count_iter:
+            x = self.model_s.extract_feat(img)
+            loss_decode, _ = self.model_s._decode_head_forward_train(x, img_metas, gt_semantic_seg=gt_semantic_seg)
+            return loss_decode
+
+        ul_data = kwargs['extra_0']
+        ul_s_img = ul_data['img']
+        ul_w_img = ul_data['ul_w_img']
+        ul_img_metas = ul_data['img_metas']
+
+        with torch.no_grad():
+            teacher_feat = self.model_t.extract_feat(ul_w_img)
+            teacher_logit = self.model_t._decode_head_forward_test(teacher_feat, ul_img_metas)
+            teacher_logit = resize(input=teacher_logit,
+                                   size=ul_w_img.shape[2:],
+                                   mode='bilinear',
+                                   align_corners=self.align_corners)
+            conf_from_teacher, pl_from_teacher = torch.max(torch.softmax(teacher_logit, axis=1), axis=1, keepdim=True)
+
+        losses = dict()
+
+        x = self.model_s.extract_feat(img)
+        x_u = self.model_s.extract_feat(ul_s_img)
+        loss_decode, _ = self.model_s._decode_head_forward_train(x, img_metas, gt_semantic_seg=gt_semantic_seg)
+        loss_decode_u, _ = self.model_s._decode_head_forward_train(x_u, ul_img_metas, gt_semantic_seg=pl_from_teacher)
```

Comment:
As I said before, please note if strong augmented image is rotated or cropped, pseudo label from teacher will not be aligned. ex) 90 rotated image - not rotated pseudo label

Reply:
Actually that step does not needed, because BranchImage is used. Please check out the code for pipelines.